module/zfs/ddt.c

```diff
@@ -1032,6 +1139,7 @@ ddt_sync_table(ddt_t *ddt, dmu_tx_t *tx, uint64_t txg)
 	memcpy(&ddt->ddt_histogram_cache, ddt->ddt_histogram,
 	    sizeof (ddt->ddt_histogram));
 	spa->spa_dedup_dspace = ~0ULL;
+	spa->spa_dedup_dsize = ~0ULL;
```

Comment:
Does `spa_dedup_dsize` need to be wiped anywhere aside of may be `ddt_load()`?  I think it may create a time window for open context between `spa_dedup_dsize` being wiped and reset when dedup is disabled for no good reason.  I don't think we need to wipe it, only update to the new up to date value during sync.

Reply:
The performance issue you raised was caused by a past regression in `ddt_get_ddt_dsize()` .  The original mechanism was that `spa_dedup_dsize` (which is a cached value) was reset after syncing the ddt and the next reader of `ddt_get_ddt_dsize()` would check if it had been reset and perform a recalculation and the cached value was good until the next txg sync.

I have restored that original behavior and made sure there were no direct consumers of `spa_dedup_dsize` and retested to confirm that we only recalculate after ddt is updated during sync.

Reply:
Thank you.  It should help with extra requests.  But it seems to me there is another issue, possibly affecting existing spa_dedup_dspace also.  I think by this time DDT ZAP is already written, but not yet synced, that means the dnode does not have the new information on used space yet.  So ddt_get_ddt_dsize() may regularly fetch and cache the value that is one TXG old.